### PR TITLE
Improve AI switching, so it doesn't get killed on switch-in

### LINKED ISSUE #2320: AI withdrawing behaviour w/o damage move, Pokemon-to-switch and next-Pokemon-sendin decisions
Hi, not a bug(?) but may be of interest for the AI.

On the master branch (17fecd5), I modded Winona1's team to the following:
1. Skarmory with {MOVE_SPIKES, MOVE_TOXIC, MOVE_WHIRLWIND, MOVE_ROOST}
2. Tropius with {MOVE_LEAF_TORNADO, MOVE_BODY_SLAM, MOVE_PROTECT, MOVE_ROOST}
3. Hawlucha with {MOVE_SUPERPOWER, MOVE_ACROBATICS, MOVE_U_TURN, MOVE_ROOST}
4. Mantine with {MOVE_CONFUSE_RAY, MOVE_WATER_PULSE, MOVE_AIR_SLASH, MOVE_ROOST}
5. Altaria with {MOVE_DRAGON_PULSE, MOVE_TOXIC, MOVE_COTTON_GUARD, MOVE_ROOST}

Winona1 has the default aiFlags AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_TRY_TO_FAINT | AI_FLAG_CHECK_VIABILITY | AI_FLAG_RISKY

Put Winona in InsideOfTruck, and something to "givemon SPECIES_RAICHU_ALOLAN, 50, ITEM_ZAP_PLATE" to the player (Raichu gets Thundershock, Tail Whip, Quick Attack and Thunderbolt), to battle immediately in a new game:

skarmorywithdraw

1. Winona switches Skarmory into Hawlucha (only one with supereff move, but is not necessarily faster + has been consistently OHKOed by Tbolt upon switchin), then Mantine (?), then Altaria then Tropius
2. Winona sends in Skarmory again (?) after the current Pokemon faints, then withdraws in the above order
3. If I replace Hawlucha's U-turn with Brick Break, Winona withdraws Skarmory into Mantine first, then Hawlucha
4. If I put Hawlucha (with U-turn) at the front and ensure it outspeeds Raichu, Hawlucha uses Superpower instead
5. If I give Skarmory Razor Wind (replacing Whirlwind), Winona does not switch Skarmory who then gets knocked out
6. If I replace Skarmory with an outspeeding Crobat with *Parting Shot*, Whirlwind, Toxic, Roost; Winona *withdraws* Crobat into Hawlucha
6a. If the Crobat has U-turn, Whirlwind, Toxic, Roost; Crobat stays and uses Toxic

If I'm not wrong:
-Switching doesn't account for the incoming Pokemon's survivability
-Switching factors in supereff of moves, but not other effects/damage/? factored by the aiFlags
-Switching is stopped by the presence of a damage move, even if the move can't get off/isn't effective/isn't picked
-Not sure about the Mantine priority (confusion? flinch?)
-Not sure about the Skarmory re-sendin
-EFFECT_HIT moves seem discouraged?

Sorry if this isn't appropriate as an issue

### PR DESCRIPTION
**Pull request by DizzyEggg**
Fixes #2320

Improves and replaces #2325

Porygon had the right idea, but his code couldn't work, because battlers were confused with party members. I modified an already existing function, so it allows to get dmg from any party mon to any party mon. 

I tested the trainer configuration in the #2320 issue and it was working, there were no dumb switches. Instead, after Skarmory fainted, I got Altaria and Tropius, the only 2 mons which survived my Raichu's Thunderbolt.

### CODE REVIEW

File: src/battle_ai_switch_items.c [hunk 1]
```diff
@@ -753,6 +754,8 @@ void AI_TrySwitchOrUseItem(void)
                     {
                         if (GetMonData(&party[monToSwitchId], MON_DATA_HP) == 0)
```
Comment (ghoulslash) [inline src/battle_ai_switch_items.c:755]: Let's use IsValidForBattle here to be more concise

File: src/battle_controller_opponent.c [hunk 2]
```diff
@@ -1702,14 +1701,14 @@ static void OpponentHandleChoosePokemon(void)
...
                 pokemonInBattle = 2;
-
             }
 
             GetAIPartyIndexes(gActiveBattler, &firstId, &lastId);
 
             for (chosenMonId = (lastId-1); chosenMonId >= firstId; chosenMonId--)
             {
                 if (GetMonData(&gEnemyParty[chosenMonId], MON_DATA_HP) != 0
+                    && GetMonData(&gEnemyParty[chosenMonId], MON_DATA_SPECIES) != SPECIES_NONE
```
Comment (ghoulslash) [inline src/battle_controller_opponent.c:1711]: IsValidForBattle

File: src/battle_ai_util.c [hunk 3]
```diff
@@ -3368,25 +3368,49 @@ bool32 ShouldUseWishAromatherapy(u8 battlerAtk, u8 battlerDef, u16 move)
...
+    if (attackerMon != NULL)
+        PokemonToBattleMon(attackerMon, &gBattleMons[battlerAtk]);
+    if (targetMon != NULL)
+        PokemonToBattleMon(targetMon, &gBattleMons[battlerDef]);
 
-    for (i = 0; i < MAX_BATTLERS_COUNT; i++)
-        gBattleMons[i] = battleMons[i];
+    for (bestDmg = 0, i = 0; i < MAX_MON_MOVES; i++)
+    {
+        if (BattlerHasAi(battlerAtk))
```
Comment (ghoulslash) [inline src/battle_ai_util.c:3400]: Can't we just check AI_PARTY->mons or does the AI info not get auto filled in?

**Review by ghoulslash — CHANGES_REQUESTED** [review]

File: src/battle_ai_util.c (hunk 3 above)
Comment (DizzyEggg) [reply src/battle_ai_util.c:3400]: the AI info doesn't get auto filled

File: src/battle_controller_opponent.c (hunk 2 above)
Comment (DizzyEggg) [reply src/battle_controller_opponent.c:1711]: Fixed

**Comment by DizzyEggg** [comment]
This is ready to review & merge

File: src/battle_controller_opponent.c [hunk 4]
```diff
@@ -1702,14 +1701,14 @@ static void OpponentHandleChoosePokemon(void)
...
                 battler2 = GetBattlerAtPosition(B_POSITION_OPPONENT_RIGHT);
                 pokemonInBattle = 2;
-
             }
 
             GetAIPartyIndexes(gActiveBattler, &firstId, &lastId);
 
             for (chosenMonId = (lastId-1); chosenMonId >= firstId; chosenMonId--)
             {
                 if (GetMonData(&gEnemyParty[chosenMonId], MON_DATA_HP) != 0
```
Comment (ghoulslash) [inline src/battle_controller_opponent.c:1710]: Looks like this could still use isValidForBattle

**Review by ghoulslash — CHANGES_REQUESTED** [review]

File: src/battle_controller_opponent.c (hunk 4 above)
Comment (DizzyEggg) [reply src/battle_controller_opponent.c]: Fixed

**Review by ghoulslash — APPROVED** [review]

**Pull request merged**